Mathlib/AlgebraicGeometry/RationalMap.lean

```diff
@@ -0,0 +1,308 @@
+  Quotient.exists_rep
+
+lemma RationalMap.exists_rep (f : X ⤏ Y) : ∃ g : X.PartialMap Y, g.toRationalMap = f :=
+  Quotient.exists_rep f
+
+lemma PartialMap.toRationalMap_eq_iff {f g : X.PartialMap Y} :
+    f.toRationalMap = g.toRationalMap ↔ f.equiv g :=
+  Quotient.eq
+
+/-- The composition of a rational map and a morphism on the right. -/
+def RationalMap.compHom (f : X ⤏ Y) (g : Y ⟶ Z) : X ⤏ Z := by
+  refine Quotient.map (PartialMap.compHom · g) ?_ f
+  intro f₁ f₂ ⟨W, hW, hWl, hWr, e⟩
+  refine ⟨W, hW, hWl, hWr, ?_⟩
+  simp only [PartialMap.restrict_domain, PartialMap.restrict_hom, PartialMap.compHom_domain,
+    PartialMap.compHom_hom] at e ⊢
+  rw [reassoc_of% e]
+
+@[simp]
+lemma RationalMap.compHom_toRationalMap (f : X.PartialMap Y) (g : Y ⟶ Z) :
+    f.toRationalMap.compHom g = (f.compHom g).toRationalMap := rfl
+
+/-- A rational map restricts to a map from `Spec K(X)`. -/
+noncomputable
+def RationalMap.fromFunctionField [IrreducibleSpace X] (f : X ⤏ Y) :
+    Spec X.functionField ⟶ Y := by
+  refine Quotient.lift PartialMap.fromFunctionField ?_ f
+  intro f g ⟨W, hW, hWl, hWr, e⟩
+  have : f.restrict W hW hWl = g.restrict W hW hWr := by ext1; rfl; rw [e]; simp
+  rw [← f.fromFunctionField_restrict hW hWl, this, g.fromFunctionField_restrict]
+
+@[simp]
+lemma RationalMap.fromFunctionField_toRationalMap [IrreducibleSpace X] (f : X.PartialMap Y) :
+    f.toRationalMap.fromFunctionField = f.fromFunctionField := rfl
+
+/--
+Given `S`-schemes `X` and `Y` such that `Y` is locally of finite type and `X` is integral,
+any `S`-morphism `Spec K(X) ⟶ Y` spreads out to a rational map from `X` to `Y`.
+-/
+noncomputable
+def RationalMap.ofFunctionField [IsIntegral X] [LocallyOfFiniteType sY]
+    (f : Spec X.functionField ⟶ Y) (h : f ≫ sY = X.fromSpecStalk _ ≫ sX) : X ⤏ Y :=
+  (PartialMap.ofFromSpecStalk sX sY f h).toRationalMap
+
+lemma RationalMap.fromFunctionField_ofFunctionField [IsIntegral X] [LocallyOfFiniteType sY]
+    (f : Spec X.functionField ⟶ Y) (h : f ≫ sY = X.fromSpecStalk _ ≫ sX) :
+    (ofFunctionField sX sY f h).fromFunctionField = f :=
+  PartialMap.fromSpecStalkOfMem_ofFromSpecStalk sX sY _ _
+
+lemma RationalMap.eq_of_fromFunctionField_eq [IsIntegral X] (f g : X.RationalMap Y)
+    (H : f.fromFunctionField = g.fromFunctionField) : f = g := by
+    obtain ⟨f, rfl⟩ := f.exists_rep
+    obtain ⟨g, rfl⟩ := g.exists_rep
+    refine PartialMap.toRationalMap_eq_iff.mpr ?_
+    exact PartialMap.equiv_of_fromSpecStalkOfMem_eq _ _ _ _ H
+
+/--
+Given `S`-schemes `X` and `Y` such that `Y` is locally of finite type and `X` is integral,
+`S`-morphisms `Spec K(X) ⟶ Y` correspond bijectively to `S`-rational maps from `X` to `Y`.
+-/
+noncomputable
+def RationalMap.equivFunctionField [IsIntegral X] [LocallyOfFiniteType sY] :
+    { f : Spec X.functionField ⟶ Y // f ≫ sY = X.fromSpecStalk _ ≫ sX} ≃
+      { f : X ⤏ Y // f.compHom sY = sX.toRationalMap } where
```

Comment:
Lets not stall this PR, but I think that our current approach to treating `S`-schemes via carrying around the additional compatibility condition everywhere won't scale.

- Maybe we should migrate to an approach to AG where `X : Scheme` is replaced by `X : C` with `[HasForget₂ C Scheme]` with a coercion from `C` to `Scheme` (and for morphisms of `C`) whenever `HasForget₂ C Scheme`. This would give us more flexibility in working with `Over S`, because everything done for schemes still applies. The additional coercions everywhere might make this horrible though.

- Or we introduce `IsOver S` and `Hom.IsOver S`. Then objects are still schemes so no coercions, but we can comfortably have compositions be automatically `Hom.IsOver S` etc. This is also close to the rest of mathlibs design, where we replaced `Type` by `Scheme`.

Reply:
Yes this is on my radar as well. I'm thinking about the latter.